### ISSUE TITLE
 You can consider aria2 support delete files?

### ISSUE THREAD

**Issue by 9E307** [aria2/aria2#728]
After using the aria2 process, we found download files, delete records, but there is no corresponding delete a file, it is desirable to be able to provide records and delete function to delete files. Thank you.

**Comment by tatsuhiro-t** [comment]
If I understand it correctly, "records" mean *.aria2 file, and you suggest that it is better to add a RPC method to delete downloaded files.  Am I correct?

**Comment by 9E307** [comment]
Can be understood, it is to allow aria2 can delete the downloaded file.

**Comment by avdhoot** [comment]
For aria2 lot web interface are available but to delete downloaded file we have ssh then delete file. It is must have feature to delete file using RPC. @mayswind

**Comment by Twilight** [comment]
If that's a remote download box, aren't you supposed to have a way to access downloaded files? like FTP or SMB share? use that.

I vote against this feature, out of the scope of a download tool.

**Comment by jiangjiangflies** [comment]
@Jimmy-Z
Jimmy-Z 
commented 2 months ago
If that's a remote download box, aren't you supposed to have a way to access downloaded files? like FTP or SMB share? use that.

I vote against this feature, out of the scope of a download tool.

too cumbersome！！！inconvenient!

**Comment by yyz1989** [comment]
+2147483647 for this feature.

If we have such an API, then we can have a downloaded file deletion checkbox in the task deletion dialog of YAAW or WebUI.

Cannot agree with the narrow perspective of someone upstairs that a download software should only focus on downloading.

**Comment by nosmokingbandit** [comment]
I'd like to request this feature as well. The api call could be as simple as `aria2.remove("token:token", "0123456789abcdef", "true")`

Being able to cancel a download through the api without being able to remove leftover half-downloaded files kind of defeats the purpose.

Until this is an option I can't fully integrate Aira2 into my application, which is a bummer because I like Aria2 and I'd love to see it gain more popularity.

**Comment by Jimmy-Z** [comment]
I suppose a "delete-files-too" option when canceling a download makes sense.

If you want it that much, this is open source software, you can always make pull requests.

BTW if you are trying to "integrate aria2 into my application", I can't see why you can't do that on your end. @nosmokingbandit

**Comment by nosmokingbandit** [comment]
Because I can't be guaranteed that my application has write access to the
download directory of a torrent client.

If I knew more C++ I'd consider diving into the code.

**Comment by no1xsyzy** [comment]
against this feature.
it is nonsense to have `rm` without `ls`, `mv`, `cp`, and other things, and aria2 can be neither better nor safer at file operations than those more professional at file operating, e.g., shells through ssh, owncloud or something like that.

Remember KISS, it is not a good idea to have a omni-tool without any customization. If there is really a requirement, just make downstream projects.

**Comment by dhanjit** [comment]
> Being able to cancel a download through the api without being able to remove leftover half-downloaded files kind of defeats the purpose

This is exactly what's required. Every bittorrent client supports this, ranging from utorrent to rtorrent. If I am removing a torrent file for example, I would definitely want the option to remove it's downloaded content.

**Comment by gokomer** [comment]
This feature is a must have for all download managers. It's a shame aria2 does not support this.

**Comment by ihipop** [comment]
any update to this issue？

**Comment by gilshallem** [comment]
please add this feature! :)

**Comment by goodandrewsoft** [comment]
Add this future finally!!!

**Comment by goodandrewsoft** [comment]
please add this feature!

**Comment by renannprado** [comment]
So sad... I believe that having the option to delete the file(s) when you want to remove the download from aria2, for example, is a pretty reasonable thing to do. 
Another example I just went through is that I stopped the download in the middle 'cause I longer wanted it and now I have files lying in the disk which sometimes are quite big files... having to delete them manually is annoying.

**Disclaimer:** I've just started to use aria2

So I propose one workaround (**that I haven't tried out yet!**) and one feature request.

**Workaround:**

Set your `--dir` to download files to `/tmp/aria2` or `/tmp`.

Before doing this, make sure that you know how/when the `/tmp` is cleaned up in your OS.

Use `--on-bt-download-complete` and `--on-download-complete` hooks to move the files that completed with success to whatever path that you want.

By doing this, at least the files that for some reason stay in the `/tmp`, eventually will be deleted from disk according to your OS' policy.

**Actual solution/feature request**

> I can move this to another issue if you want 

Implement hooks like `--on-bt-download-removed` and `--on-download-removed`. With such a hook, one could implement a simple shellscript that would remove the files after they are removed from aria2. Should such a hook fail (i.e. exit status != 0) then the download shouldn't be removed from aria2.

@Jimmy-Z @no1xsyzy what do you think?

**Comment by Jimmy-Z** [comment]
@renannprado Funny thing though, I'm not a developer of aria2...

That said, it might be possible already in current aria2, I'm not sure which one will be called, `--on-download-complete` or `error`, you can simply try, then call the RPC interface [tellStatus](http://aria2.github.io/manual/en/html/aria2c.html#aria2.tellStatus), if it's `removed`, delete related files.

Hooks are fun, I wrote a complete hook to send me push notifications, then I disabled it since it's too annoying.

BTW, exercise extreme caution when writing scripts that actually delete files.

**Comment by Jimmy-Z** [comment]
It's still a bit inconvenient though, some times the user might want to cancel the download but keep the incomplete file(s), there is no way to pass that information to the hook.

**Comment by renannprado** [comment]
@Jimmy-Z I only mentioned you 'cause you didn't seem convinced that having aria2 to delete files is not a good feature to have.

I'll give a try to the existing hooks then.

> It's still a bit inconvenient though, some times the user might want to cancel the download but keep the incomplete file(s), there is no way to pass that information to the hook.

This is a valid point, but at same time it seems to me that keeping half-way downloaded files is not something that most of the people will care for. Even then, I can imagine it would be possible to add parameter to the method that removes the metadata so that it can let the hook know whether or not the user wanted the files to be deleted.

**Comment by HyperClockUp** [comment]
It's sad to hear that there is still no way to automaticlly delete files at 2021. Sometimes when I want to delete the download directory, there is always a file named *.aria2_temp is used except after you waiting for a while. So the best solution is to ADD THE FEATURE to remove all the useless files automatically.

**Comment by yunyuyuan** [comment]
Now is 2023,waiting for this feature...
I think aria2 can provide a download cache folder(like other download manager),once the download is complete,move the file to `~/Downloads`.And give user a option to remove uncompleted cache files.

**Comment by jjling2011** [comment]
It's 2026 now. We can use AI to write a script to fulfill this requirement!

The script logic is as follows:
 1. Use JSON-RPC to fetch download directories of all active and paused tasks (name as "task_dirs").
 2. Walk throught download dir and collect subdirectories that are not in "task_dirs" (name as "del_dirs").
 3. recursively delete those "del_dirs".

[cleanup-download-folder-example.py](https://gist.github.com/jjling2011/61e513d7fadb3d48dee132f43f753ae1)
Warning: Do not run this script. Ask AI to generate your own version.
